Generalize objects

xiangao1 opened the pull request:
## Summary/Motivation:
Generalize the double-loop simulation framework for energy systems with the grid.

## Changes proposed in this PR:
- Generalize existing MIP models into objects
- Generalize required plugins to Prescient

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

xiangao1 committed:
intialize files

xiangao1 committed:
initialize test files

xiangao1 committed:
move old code here

xiangao1 commented:
Hi @adowling2, I have moved the old code I had into this repo now. Would you mind looking at it? Thank you!

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+    Returns:
+        df: a dataframe that has the information we specified.
+    '''
+
```
Consider performing checks here to see if the data frame contains the necessary fields.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+
+class DAM_thermal_bidding:
+
+    n_scenario = 10
```
Move to self, put in __init__

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+        '''
+
+        # read the price forecasts
+        self.price_forecasts = pd.read_csv(price_forecast_file,index_col = ['Date','Hour'])
```
Despite consideration: what do you think about moving this to be helper functions or a separate class? The input here would be a pandas dataframe. I think this would facilitate easier interfacing with Prescient.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+        '''
+
+        # read data
+        gen_params = pd.read_csv(rts_gmlc_data_dir + 'gen.csv')
```
Is this RTS-GMLC specific? Would it make more sense to have an RTS-GMLC data class that handles all of this?

adowling2 replied:
The bidder here would just take pandas dataframes as input.

adowling2 replied:
Or you could refine a `gen_params` data class.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+        self.bids_result_list = []
+
+    @staticmethod
+    def assemble_model_data(generator_names, rts_gmlc_data_dir, **kwargs):
```
Does it make sense for generate_names and any RTS-GMLC data to live in self? Then this would not be a static method.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+        return m
+
+    @staticmethod
+    def _update_UT_DT(m,implemented_shut_down, implemented_start_up):
```
@jsiirola What are your thoughts on leaving this as a static method with `m` as an input or having `m` live in `self` and making this a normal class method?

adowling2 replied:
Same comment applies to multiple functions.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+                m.DAM_price[t,i] = forecasts[t,i]
+
+        # solve the model
+        solver = pyo.SolverFactory('gurobi')
```
Make solver live in self. That way the user can define a custom solver. If `solver=None` in `__init__`, then use Gurobi as the default. We'll need to use CBC for automated testing.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+                        yield m.gen_pow[j,h,k[0]] == m.gen_pow[j,h,k[1]]
+        m.NAC_SS_cons = ConstraintList(rule = NAC_list_rules)
+
+        return m
```
This function returns `m`, but similar functions do not. Not sure if this is intentional.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,774 @@
+        # record the bids
+        self._record_bids(date,bids)
+
+        return bids
```
How is `bids` here different from `bids_result_list` created in `_record_bids`? What format of bids does Prescient need?

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+
+    return cnt
+
+def assemble_model_data(generator_names, \
```
What do you think about, after generalizing, making one of the examples this thermal generator with optional storage? I am okay with RTS-GMLC data living in that example.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+
+    model_data['Min Load Cost'] = dict(zip(generator_names,gen_params['HR_avg_0']/1000 * gen_params['Fuel Price $/MMBTU'] * gen_params['PMin MW']))
+
+    if storage_size_ratio is None:
```
Add a sentence comment after each part of this else/if block to explain what is happening.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+
+    return m
+
+def _add_tracking_constraints_and_objectives(m, ramping_weight = 20):
```
Should this part get abstracted to the tracker object?

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+    return
+
+def build_tracking_model(model_data,plan_horizon,ramping_weight = 20,segment_number=4,model_mode = '1a'):
+
```
Missing doc string.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+    return m
+
+def switch_model_mode(m,mode = '1a'):
+
```
Missing doc string.

adowling2 replied:
I recommend making the modes more descriptive. `no-storage-market-interactions` instead of `1a`.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+
+    return m
+
+def record_results(m, time, market_signals, **kwargs):
```
Nice. This gets generalized.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+
+    return
+
+def pass_schedule_to_track(m,market_signals,implement_steps = 1):
```
Nice. This gets generalized.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+    return
+
+def update_model(m,last_implemented_time_step = 0):
+
```
Missing doc strings.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+    return solver_status
+
+def _update_power(m,last_implemented_time_step = 0):
+
```
Missing doc strings.

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_storage.py:
```diff
@@ -0,0 +1,1069 @@
+    return
+
+def solve_relaxed_model(m,solver):
+    '''
```
Should this be part of the generalized workflow? My gut says yes, as a PBC will be a standard feature in most of the IES models.

adowling2 commented on idaes/apps/grid_integration/prescient_plugin.py:
```diff
@@ -0,0 +1,524 @@
+    '''
+
+    # initialize the model class
+    bidder = DAM_thermal_bidding(rts_gmlc_data_dir = options.rts_gmlc_data_dir, \
```
How should we generalize this to support multiple user-defined IES? Should the user-defined classes be an input to this function?

adowling2 commented on idaes/apps/grid_integration/prescient_plugin.py:
```diff
@@ -0,0 +1,524 @@
+    return
+prescient.plugins.register_initialization_callback(initialize_bidding_object)
+
+def initialize_tracking_object(options, simulator):
```
Same comment as above.

adowling2 commented on idaes/apps/grid_integration/prescient_plugin.py:
```diff
@@ -0,0 +1,524 @@
+        None
+    '''
+
+    gen_name = options.bidding_generator
```
Think about the case where we replace a node in RTS-GMLC with a full IDAES IES model. We might not use the RTS-GMLC data to define the bidding and tracker. Right now, your bidder and tracker uses the RTS-GMLC data directly.

adowling2 commented on idaes/apps/grid_integration/prescient_plugin.py:
```diff
@@ -0,0 +1,524 @@
+        market_signals: the market signals to be tracked.
+    '''
+
+    ## TODO: make this in a for loop (imagine we have a list of bidding generators)
```
Agreed. Support for multiple bidding generators is a good future TODO, but not critical for the initial release.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+import os
+
+def get_data_given(df, bus=None, date=None, hour=None, generator=None, fuel_type=None):
+
```
Thoughts on an RTS-GMLC class that handles data manipulation?

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+
+class DAM_thermal_tracking:
+
+    def __init__(self,rts_gmlc_data_dir,tracking_horizon = 4, generators = None):
```
Add solver as an optional input.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        self.projection = {}
+
+    @staticmethod
+    def assemble_model_data(generator_names, rts_gmlc_data_dir, **kwargs):
```
Consider moving this to an RTS-GMLC specific class.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        return m
+
+    @staticmethod
+    def _add_UT_DT_constraints(m):
```
How much of this is general versus model-specific?

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        return
+
+    def build_tracking_model(self, generators, track_horizon):
+
```
Doc string.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+                m.power_dispatch[unit,t] = market_signals[unit][t]
+
+        if solver is None:
+            solver = pyo.SolverFactory('gurobi')
```
Put in self?

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        return
+
+    def _record_daily_stats(self,m,last_implemented_time_step=0):
+
```
How should this be abstracted?

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        return
+
+    def _record_results(self, m, market_signals, date = None, hour = None, **kwargs):
+
```
Nice. This looks generalized already.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+
+    @staticmethod
+    def _build_model(model_data,plan_horizon,segment_number=4):
+
```
How can we reduce the amount of duplicate code for this example when defining the bidder model and tracker model objects?

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+
+    @staticmethod
+    def _add_tracking_constraints_and_objectives(m, ramping_weight = 20):
+
```
Doc string.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        return
+
+    def build_tracking_model(self,plan_horizon,ramping_weight = 20,segment_number=4,model_mode = '1a'):
+
```
Doc string.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+
+    @staticmethod
+    def switch_model_mode(m,mode = '1a'):
+
```
This appears to be duplicate with the bidder. Happy to talk about ways to reduce redundant code.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1747 @@
+        return m
+
+    @staticmethod
+    def record_results(m, time, market_signals, **kwargs):
```
Already appears to be generalized. How is this different from the bidder?

adowling2 reviewed (commented):
@xiangao1 Nice prototype. Below are some comments in anticipation of your generalization.

xiangao1 committed:
add thermal generator class

xiangao1 committed:
update the doc strings in the thermal generator object

xiangao1 committed:
add the rts gmlc dataset for testing for now

xiangao1 commented:
@adowling2 @jsiirola I have added the thermal generator model object in the example folder `thermal_generator.py`.  As we have discussed this object has methods to 

1. create a Pyomo model: `build_thermal_generator_model`
2. advance the time step: `update_model`
3. record the results: `record_results`

For convenience, I added the RTS-GMLC dataset for testing in the folder as well. 

Following the timeline we had, I plan to implement the thermal generator + storage model object tomorrow. That object will be built based on this thermal generator object so we don't have to maintain duplicate code.

After that, I plan to generalize `bidder` and `tracker` objects to be IES model agnostic. In that way, the users will only have to provide the power output and production cost variable names and indices in their model.

adowling2 commented:
@xiangao1 Great. On Friday, we can also discuss moving the static method `assemble_model_data` into an RTS-GMLC class. This would mean ThermalGenerator takes model_data as an input. Would could then test ThermalGenerator without the RTS-GMLC data.

jsiirola commented:
One quick note: you may want to retarget this PR to not merge into  `main@xiangao1/idaes-pse`, but to merge into a separate branch (e.g., `double-loop@xiangao1/idaes-pse`).  Merging into main on your fork will start to get very confusing and make it harder for you to keep your branches up-to-date with the upstream main branch on IDAES/idaes-pse.

adowling2 commented:
@jsiirola Great suggestion. If we cannot get it working a few minutes, we might ask you or Ben to show us tomorrow on the call.

xiangao1 commented:
@adowling2 @jsiirola Now it will be merged into `double-loop@xiangao1/idaes-pse`. Thanks for the suggestion!

xiangao1 committed:
outline for the ies model object

xiangao1 committed:
add the var and constraints for ies

xiangao1 committed:
add scenario indices

xiangao1 committed:
add record result method for ies model object

xiangao1 committed:
add update model method for ies

xiangao1 committed:
add ies mode switch method

xiangao1 committed:
add doc strings

xiangao1 commented:
@adowling2 @jsiirola I have added the IES model object in the example folder thermal_generator_storage.py. As we have discussed this object has methods to

create a Pyomo model: `build_thermal_generator_storage_IES_model`
advance the time step: `update_model`
record the results: `record_results`

So right now this file reuses code from the thermal generator and avoids duplicate code. 

I will prepare some questions for today's meeting.

xiangao1 committed:
add object property for power, cost and indices

xiangao1 committed:
associate costs with weights

xiangao1 committed:
finished adding tracking constraints and objective  to model object

xiangao1 committed:
test with thermal generator

xiangao1 committed:
revise model update functions

xiangao1 committed:
finish generalizing the tracker object

xiangao1 committed:
add doc strings

xiangao1 committed:
make progress on bidder

xiangao1 committed:
finish bidding constraint

xiangao1 committed:
successfully add bidding constraints and objective

xiangao1 committed:
first draft of bidder

xiangao1 committed:
add object properties

xiangao1 committed:
test and debug bidder with the thermal generator model

xiangao1 committed:
add doc strings

xiangao1 commented:
@adowling2 @jsiirola I have generalized the `tracker` and `bidder` objects over the last week. 

As we have discussed the `Tracker` object in the `tracker.py` file has methods to:
- add constraints and objective function for the tracking problem: `formulate_tracking_problem()`
- track the market dispatches: update_model: `track_market_dispatch()`
- record the results:`record_results()`

As for the `Bidder` object in the `bidder.py` file, it has methods to:
- add constraints and objective function for the bidding problem: `formulate_bidding_problem()`
- compute the optimal bids given LMP forecasts: update_model: `compute_bids()`
- record the bids:`record_bids()`

Because the goal of the generalization is to take arbitrary energy system models and formulate the tracking/bidding problem for them, as tracker/bidder we do not necessarily know the index sets in the model, i.e., number of index sets, the how the variables are indexed. In order to tackle this problem, I came up with a recursive way to do this. I am going to use the `_add_tracking_dispatch_constraints()` method to explain.

1. This constraint says the a certain generation unit `g` in the model needs to have the power output at the market dispatch at a certain time `t`. So I assume the model must have a set for units and time. When we iterate over them, we need to 'label' it.
2. As for the other sets the model might have, we just need to recursively traverse. When we have traversed all the sets the model has, we add the constraints.

And the `bidder` object uses the same idea to add constraints and objective. Let me know if you have any comments. Thank you!

This week I am going to work on generalize the object that talks to Prescient, `tracker`, and `bidder`.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+        bidding_set = []
+
+        model_sets = self.bidding_model_object.indices
```
@jsiirola @xiangao1 What do you think about adding an assert statement here or similar to ensure `indices` is as expected?

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+        Returns:
+            bidding_set: a list that contains the necessary Pyomo set objects.
+        '''
```
Perhaps include an example in the doc string. If `self.bidding_modeling_objects.indices` contains ... then bidding_set will contain ...

Alternately, we could write a test for this.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+            None
+        '''
+
+        # add params to the model
```
I recommend expanding this comment and explaining the * syntax.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+        for s, n in model_sets.items():
+            if n == 'LMP Scenarios':
+                modified_model_sets[tuple(combinations(s,2))] = n
```
Add a few comments to explain this

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+        # declare a constraint list
+        self.model.bidding_constraints = pyo.ConstraintList()
+
+        # dfs
```
Define dfs

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+        modified_set_ls = list(modified_model_sets.keys())
+
+        # unpack
+        power_output = self.bidding_model_object.power_output
```
@jsiirola @xiangao1 What do you think about adding 10 to 20 lines of code in `__init__` to check all of the inputs to this class. For example, check that the model has a `power_output` and an `energy_price` attribute.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+             # traveled all the sets
+             # add the bidding constraints
+             if set_idx == len(modified_set_ls):
+                 self.model.bidding_constraints.add((power_output[tuple(indices_1)] - \
```
Add a comment explaining this constraint.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+             for idx in modified_set_ls[set_idx]:
+
+                 if modified_model_sets[modified_set_ls[set_idx]] == 'LMP Scenarios':
```
Add comments explaining this.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+                 # recursion
+                 dfs_add_bidding_constraints(set_idx + 1, indices_1, indices_2, bidding_indices_1, bidding_indices_2)
+
+                 # backtrack
```
I would add a block comment in this method listing out the main steps (recursion, backtrack, etc.). This code looks nice but is difficult to understand line by line quickly. Comments can help make it faster to "digest".

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+        # dfs
+        def dfs_add_bidding_constraints(set_idx, indices_1, indices_2, bidding_indices_1, bidding_indices_2):
+
```
Doc string

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+        # declare an empty objective
+        self.model.obj = pyo.Objective(expr = 0, sense = pyo.maximize)
+
+        # unpack things
```
unpack indices?

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+            for idx2 in self.bidding_set[1]:
+                for idx3 in self.bidding_set[2]:
+
+                    aux_dict[model_sets[self.bidding_set[0]]] = idx1
```
Add comments.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+            model_sets_inverse[n] = s
+
+        # unpack things
+        model_sets = self.bidding_model_object.indices
```
Redefined model_sets? Perhaps this line can be deleted.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+        # unpack things
+        model_sets = self.bidding_model_object.indices
+        set_ls = list(model_sets.keys())
```
Would it make sense to have an `unpack` method that returned `set_ls`, `power_output`, and `energy_price`?

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+            # traveled all the sets, and now we can add the objective
+            if set_idx == len(set_ls):
+                power = round(pyo.value(power_output[tuple(indices)]),2)
```
Is the rounding important or just for ease of reading? My preference is to set `self.bid_rounding_threshold = 2` in the `__init__` method and then use it here. That way it is easier to change. Or we can skip rounding and do it later.

adowling2 replied:
Or if rounding is important, add a comment that explains why to keep this at 2.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+
+        for t in model_sets_inverse['Time']:
+            for gen in model_sets_inverse['Generators']:
+
```
The comments in this section of nested for loops are really helpful. My suggestion is to comment the rest of the class with similar detail.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,1951 @@
+
+        # unpack things
+        model_sets = self.tracking_model_object.indices
+        set_ls = list(model_sets.keys())
```
Does it make sense to define an `unpack` method in each class. Looks like this is used several places.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+        bids = self._assemble_bids()
+        self.record_bids(bids, date = date, hour = hour)
+
+        ## TODO: when to update bidding model??
```
Happy to discuss this more during a meeting.

adowling2 commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,1160 @@
+        # get the sets needed for the tracking problem
+        self.bidding_set = self._get_bidding_sets()
+
+        self._add_bidding_params()
```
Is price data needed to define the bidding problem? How does the user specify the horizon length?

xiangao1 replied:
The bidder/tracker 'inherit' the horizon length from the model object.

adowling2 reviewed (commented):
@xiangao1 Below are my comments. I like the generalization for traversing model indices. My biggest recommendation is to add comments to explain the logic and steps. I can then re-review the logic more carefully.

xiangao1 committed:
add more doc strings

xiangao1 committed:
add comments

xiangao1 committed:
outline the object

xiangao1 committed:
fixing typos

xiangao1 committed:
fix indentation

xiangao1 committed:
build thermal genearator with only time index

xiangao1 committed:
fix the helper functions

xiangao1 committed:
build IES with only time indices

xiangao1 committed:
fix the helper functions in IES object

xiangao1 committed:
update doc stringsa

xiangao1 committed:
refactor tracker

xiangao1 committed:
remove old code

xiangao1 committed:
create block in model object

xiangao1 committed:
tracker take blocks

xiangao1 committed:
refactor thermal gen object

xiangao1 committed:
refactor tracker

xiangao1 committed:
some progress

xiangao1 committed:
bidder works

xiangao1 committed:
small changes to adapt tracker and bidder

xiangao1 committed:
add write result functions

xiangao1 committed:
add functions to get the latest power output

xiangao1 committed:
get rid of simulator.data_manager.extension

xiangao1 committed:
finish projection capabilities

xiangao1 committed:
fix typos

xiangao1 committed:
plug in a price forecaster object in bidder

xiangao1 committed:
add and test a very simple forecaster

xiangao1 committed:
add more price forecasts

xiangao1 committed:
debug

xiangao1 committed:
simulate with prescient

xiangao1 committed:
enable after simulation plugin

xiangao1 committed:
adapt the options to the newest prescient

xiangao1 committed:
renaming finalization

xiangao1 committed:
add doc strings

xiangao1 committed:
add doc strings for tracker methods

xiangao1 committed:
add doc strings in bidder

xiangao1 committed:
remove old code

xiangao1 committed:
clean up imports

xiangao1 committed:
fix typo

xiangao1 committed:
update doc strings

xiangao1 committed:
more doc strings

adowling2 commented:
@xiangao1 Based on a conversation with @jsiirola and @jalving today, it would be very helpful if the interface for tracker (and bidder, but less important) could be used standalone from Prescient. Specifically for the tracker, we want to:
- Give the input data (signal to track) as a numpy array, list, or similar data type
- Run the tracker in a rolling horizon without Prescient. This might require writing an external rolling horizon wrapper, which would not be too difficult.

We implicitly discussed this on Tuesday regarding testing each component separately.

@jalving From the bidder and tracker, we could make a price_taker class that does the analysis we discussed today.

xiangao1 commented:
Hi @adowling2 @jsiirola @bknueven @jalving @nicolepcortes. I want to give you an update on my progress. As we have discussed several times, the double loop mainly needs 4 objects/classes, i.e., model object, tracker, bidder, and interface object. I have implemented these capabilities and below is a brief summary.

`Tracker`: in `tracker.py`. This class takes a model object class. It adds tracking objective function and constraints to the underlying model. When being called it solves the tracking problem.

`Bidder`: in `bidder.py`. This class takes a model object class and a price forecaster. It adds tracking objective function and constraints to the underlying model. When being called the price forecaster forecasts and then the bidder solves the bidding problem. 

`DoubleLoopCoordinator`: in `prescient_plugin.py`. This class contains the functionalities to coordinate bidder, tracker, and Prescient. When being initialized, it registers these methods in Prescient's plugin system. 

In the examples folder, `thermal_generator_prescient_plugin.py` shows an example of how someone can use the double loop. Basically, the user needs to initialize bidder, tracker, and coordinator and then tell Prescient where this py file is (via command-line option e.g., --plugin=thermal_generator_prescient_plugin). 

I would really appreciate your feedback, especially on how to test these classes. Thanks!

xiangao1 commented:
> @xiangao1 Based on a conversation with @jsiirola and @jalving today, it would be very helpful if the interface for tracker (and bidder, but less important) could be used standalone from Prescient. Specifically for the tracker, we want to:
> 
> * Give the input data (signal to track) as a numpy array, list, or similar data type
> * Run the tracker in a rolling horizon without Prescient. This might require writing an external rolling horizon wrapper, which would not be too difficult.
> 
> We implicitly discussed this on Tuesday regarding testing each component separately.
> 
> @jalving From the bidder and tracker, we could make a price_taker class that does the analysis we discussed today.

@adowling2 Yes, they can be used without Prescient. We just need another object to dispatch load signals in rolling horizon.

bknueven commented:
> @adowling2 Yes, they can be used without Prescient. We just need another object to dispatch load signals in rolling horizon.

I should mention that the short-cut simulator for Prescient it basically done (available [here](https://github.com/grid-parity-exchange/Prescient/tree/shortcut_simulator)), but it's waiting on a PR in Egret and an subsequent PR in Prescient before I open a PR for it. It's uncomplicated, so it should be part of Prescient in the next week or so.

xiangao1 committed:
getting started to test tracker

xiangao1 committed:
fix typo in file name

xiangao1 committed:
add test for tracker

adowling2 commented on idaes/apps/grid_integration/examples/thermal_generator_prescient_plugin.py:
```diff
@@ -0,0 +1,54 @@
+
+# create forecaster
+price_forecasts_df = pd.read_csv('lmp_forecasts_concat.csv')
+forecaster = WhiteNoiseForecaster(price_forecasts_df = price_forecasts_df)
```
Add support to seed the random number generator to make this reproducible?

adowling2 commented on idaes/apps/grid_integration/forecaster.py:
```diff
@@ -0,0 +1,27 @@
+        self._rename_columns()
+
+    def _rename_columns(self):
+        col_mapping = {"Scenario {}".format(i): i - 1 for i in range(1, 11)}
```
Why is range(1, 11) hard coded?

adowling2 commented on idaes/apps/grid_integration/forecaster.py:
```diff
@@ -0,0 +1,27 @@
+import pandas as pd
+
+class WhiteNoiseForecaster:
```
Needs doc string. Does this read in forecasts from file or add white noise? If it just reads from file, rename the class accordingly.

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,261 @@
+        '''
+
+        # create an instance
+        self.tracking_model_object = tracking_model_class(**kwarg)
```
Why not pass in tracking_model_object?

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,261 @@
+
+        # add flowsheet to model
+        self.model = pyo.ConcreteModel()
+        self.model.fs = pyo.Block()
```
self.model.fs = self.tracking_model_object.create_model()

adowling2 replied:
This avoids need to call `transfer_attributes_form`

adowling2 replied:
Bidder can do the same thing. Rule calls create model.

adowling2 replied:
self.model.fs[t] = self.tracking_model_object.create_model()

adowling2 replied:
See discussion from John.

adowling2 replied:
Suggestion from @jsiirola:
```
model = ConcreteModel()
model.fs = Block(model.TIME)
model.fs[1] = self.tracker_model.create_model()
# ^^^ this works, but Pyomo will call transfer_attributes_from under the hood

# ...alternatively:

self.tracker_model.populate_{something}(model.fs[1])

# If you want to take an arbitrary user model

class Wrapper():
    def populate_model(b):
        b.user_model = create_user_model()
```

adowling2 replied:
@jalving See this

adowling2 replied:
@jsiirola Is going to draft up a clean implementation for @jalving and possibly @xiangao1

adowling2 commented on idaes/apps/grid_integration/tracker.py:
```diff
@@ -0,0 +1,261 @@
+        self.model.fs.transfer_attributes_from(b)
+
+        # get the power output
+        power_output_name = self.tracking_model_object.power_output
```
This should throw a specific error if tracking_model_object does not have `power_output`.

adowling2 replied:
Or perhaps the model object should throw specific errors if `create_model`, etc. are not properly defined.

adowling2 reviewed (commented):
Notes from meeting today:
1. First get regression tests working
2. Add exceptions for model object setup wrong
3. Add unit tests to verify exceptions are thrown
4. Get help from @jsiirola to test both expectation type and exception method.

xiangao1 committed:
adapt to populate model

xiangao1 committed:
update tracker test to popluate model

xiangao1 committed:
fix a typo

xiangao1 committed:
renaming PlaceHolderForecaster

xiangao1 committed:
update the example

xiangao1 committed:
add doc string for the placeholder forecaster

xiangao1 committed:
pass more args when compute bids

xiangao1 committed:
start testing bidder

xiangao1 committed:
tests for bidder pass

xiangao1 committed:
minor

xiangao1 committed:
Merge branch 'generalize-objects' of https://github.com/xiangao1/idaes-pse into generalize-objects

xiangao1 committed:
rearrange functions for testing

xiangao1 committed:
rename activate_DA_bids

xiangao1 committed:
fix a bug

xiangao1 committed:
rename generalized plugin to coordinator

bknueven commented:
@xiangao1, when you get a minute can you update this for the new Prescient plugin interface? [This](https://github.com/grid-parity-exchange/Prescient/tree/aeb40178ef4b4f7007f65c0f2e2206821ff8ebda) is the last Prescient commit the current code will work with.

You can find a new example plugin here: https://github.com/grid-parity-exchange/Prescient/blob/main/tests/simulator_tests/test_cases/test_plugin.py.

xiangao1 committed:
start testing coordinator

xiangao1 committed:
tests passed

xiangao1 commented:
https://github.com/xiangao1/idaes-pse/blob/c2fad19c956d4678eeb0c25a0f5f1967e3abfd17/idaes/apps/grid_integration/tracker.py#L141
How to refactor to support resolve with different weights?

xiangao1 committed:
register functions using context

xiangao1 committed:
comment out traceback and run-deterministic-ruc

xiangao1 committed:
remove outdated customized command line options

xiangao1 committed:
register customized commandline options using pyomo config

xiangao1 commented:
Hi @bknueven I just created this PR #2 to adapt to the new prescient interface. Would you mind reviewing it? I really appreciate your feedback! Thanks!

xiangao1 committed:
store plugin config as property

xiangao1 committed:
using prescient python api

xiangao1 committed:
remove the txt config file

xiangao1 committed:
Merge pull request #2 from xiangao1/adapt-new-prescient-plugin-system

Adapt to new prescient plugin system

xiangao1 committed:
start testing tracker

xiangao1 committed:
add functions to test missing methods and attr

xiangao1 committed:
use regex to check error message

xiangao1 committed:
simplify the code

minor

xiangao1 committed:
add the same tests for bidder

xiangao1 committed:
add checker and tests for inputs to the tracker

xiangao1 committed:
check inputs into bidder and test

xiangao1 committed:
add doc strings

Commit:
working multiperiod implementation

Commit:
fix up comments

Commit:
update doc strings

Commit:
fix spacing

Commit:
fix more spacing

xiangao1 committed:
fix the imports

xiangao1 committed:
move number of segement as a class attribute

adowling2 committed:
Update thermal_generator.py

Added comment to explain 4 segments

Commit:
add jupyter notebook for rankine multi-period

Commit:
update notebook, allow 'build_multi_period_model' with no arguments

Commit:
add rankine folder to idaes apps

jalving committed:
Merge pull request #3 from xiangao1/generalize-objects

Generalize objects

Commit:
wip rankine double loop

Commit:
update rankine double loop

Commit:
bidder and tracker solve

Commit:
got double loop results

Commit:
update double loop for merge

Commit:
move multiperiod folder

Commit:
remove old multiperiod

Commit:
remove dev file

Commit:
more cleanup

Commit:
update prescient run script

jalving committed:
Delete idaes/apps/grid_integration/examples/multiperiod_rankine/bidding_plugin_test_multiperiod_rankine directory

Commit:
update run script

Commit:
Merge branch 'double-loop-rankine' of github.com:jalving/idaes-pse into double-loop-rankine

xiangao1 committed:
Merge pull request #4 from jalving/double-loop-rankine

Double loop rankine

xiangao1 committed:
update the doc strings

xiangao1 committed:
add sphinx doc files

xiangao1 committed:
Merge branch 'tests-for-objects' of https://github.com/xiangao1/idaes-pse into tests-for-objects

xiangao1 committed:
fix conflict

xiangao1 committed:
remove extra test file

xiangao1 committed:
add more comments on code

xiangao1 committed:
formmating code

xiangao1 committed:
add function to check convexity of the bids and corresponding tests

add doc strings

xiangao1 committed:
reformat the tests with black

xiangao1 committed:
improve the documents

xiangao1 committed:
add more figures

xiangao1 commented:
Hi @adowling2, I have added tests to check the convexity of the bids. Would you mind taking a look at them? The tests are here: grid_integration/tests/test_bidder.py. Thanks!

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+
+
+class TestingForecaster:
+    def __init__(self, horizon, n_sample):
```
Add a 1 sentence comment explaining what this class does.

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    def __init__(self, missing_method=None, missing_attr=None):
+
```
Add a few sentence comment explaining what __init__ does. My guess is this creates an object with the dummy functions specified in `method_dict` and attributes specified in `attr_dict`, but it would be best to be explicit about this in the comments.

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+    solver = pyo.SolverFactory("cbc")
+    forecaster = TestingForecaster(horizon=horizon, n_sample=n_scenario)
+
+    method_list = ["populate_model", "update_model"]
```
Add comment such as "By definition, the model object should contain these methods"

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+    solver = pyo.SolverFactory("cbc")
+    forecaster = TestingForecaster(horizon=horizon, n_sample=n_scenario)
+
+    attr_list = ["power_output", "total_cost", "generator", "pmin", "default_bids"]
```
By definition, the Forecaster object should contain these attributes

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+@pytest.mark.unit
+def test_solver_checker():
+
+    solver = pyo.SolverFactory("cbc")
```
Is this needed?

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+        for t in range(horizon)
+    }
+
+    assert expected_bids == bids
```
Did we want to use `==` or did we discuss an alternate utility provided in Pyomo (or IDAES) for comparing dictionary-like objects within a specified numeric accuracy?

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+    # expect default bids
+    shift = 1
+    fixed_forecast = marginal_cost - shift
+    bids = bidder_object.compute_bids(date=date, hour=None, prediction=fixed_forecast)
```
What happens if an invalid date is provided to `compute_bids`? Do we need a test for this resulting error message?

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+        pre_cost = 0
+
+        for p in default_bids:
+            if p == pmax:
```
Add a 1 sentence comment explaining where this formula came from

adowling2 commented on idaes/apps/grid_integration/tests/test_bidder.py:
```diff
@@ -0,0 +1,207 @@
+            pre_p = p
+            pre_cost = expected_bids[t][gen][p]
+
+    assert expected_bids == bids
```
See prior comment about using `==` or something else

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+
+    # test if it assembles correct signals within a day
+    hour = 10
+    expected_signal = [(hour + 1 + i) * 10 for i in range(sced_horizon)]
```
Add a sentence explaining the expected signal is a constant value

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+def test_assemble_project_tracking_signal(coordinator_object):
+
+    gen_name = coordinator_object.bidder.generator
+    current_ruc_dispatch = {(gen_name, t): (t + 1) * 10 for t in range(24)}
```
Replace instances of `10` with a variable named `constant_price`.

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+@pytest.mark.unit
+def test_assemble_project_tracking_signal(coordinator_object):
+
+    gen_name = coordinator_object.bidder.generator
```
Overall need to explain the `current_ruc_dispatch` signal and expected signals are all a constant price for this unit test

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+def test_assemble_sced_tracking_market_signals(coordinator_object):
+
+    gen_name = coordinator_object.bidder.generator
+    sced_dispatch = [20] * tracking_horizon
```
Add comment explaining `sced_dispatch` is a constant signal for this unit test

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+    gen_name = coordinator_object.bidder.generator
+    sced_dispatch = [20] * tracking_horizon
+    sced_horizon = tracking_horizon
+    current_ruc_dispatch = {(gen_name, t): (t + 1) * 10 for t in range(24)}
```
Add comment to explain the shape of this signal

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+    # test case 1: no ruc signals from next day
+    hour = 10
+    next_ruc_dispatch = None
+    expected_signal = [sced_dispatch[0]] + [
```
Add comment to explain the shape of this signal

adowling2 commented on idaes/apps/grid_integration/tests/test_coordinator.py:
```diff
@@ -0,0 +1,141 @@
+
+    # test case 3: with ruc signals, between 2 days
+    hour = 23
+    next_ruc_dispatch = {(gen_name, t): (t + 1) * 10 for t in range(24)}
```
Add comments to explain the shapes of these signals

adowling2 commented on idaes/apps/grid_integration/tests/test_tracker.py:
```diff
@@ -0,0 +1,397 @@
+    A class for testing missing methods and attributes.
+    """
+
+    method_list = [
```
Same comments as `test_bidded.py` for this function

adowling2 commented on idaes/apps/grid_integration/tests/test_tracker.py:
```diff
@@ -0,0 +1,397 @@
+    Simple model object for testing.
+    """
+
+    pmin = 20.00
```
Should these be in the `init` method?

adowling2 reviewed (commented):
Added comments on the tests. These look great!

jalving commented on idaes/apps/grid_integration/bidder.py:
```diff
@@ -0,0 +1,504 @@
+        # check whether the marginal costs are sorted <=> convex
+        idx = 0
+        while idx < len(marginal_cost) - 1:
+            if marginal_cost[idx] > marginal_cost[idx + 1]:
```
This check can fail due to numerical comparison. This was causing issues for me when the marginal cost was supposed to be constant, but the the true values were different by 1e-12. We should round within precision for this check.

xiangao1 committed:
improve the docs with alex advices

xiangao1 committed:
address comments from alex

xiangao1 committed:
using pyomo assertStructuredAlmostEqual

xiangao1 committed:
Merge pull request #5 from xiangao1/improve-docs

improve the documents
